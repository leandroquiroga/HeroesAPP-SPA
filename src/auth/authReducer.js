import { types } from "../types/type";



export const authReducer = (state = {}, action) => {
  const { type, payload } = action;
  
  switch (type) {
    case types.login:
      return {
        ...payload,
        logged: true,
      };
    case types.logout:
      return {
        logged: false,
      };
    
    default:
      return state;
  }
}