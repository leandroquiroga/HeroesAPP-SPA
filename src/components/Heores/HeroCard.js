import React from 'react';
import { Link } from 'react-router-dom';

export const HeroCard = ({
  id,
  superhero,
  publisher, 
  alter_ego,
  first_appearance,
  characters,
}) => {
  const imageSRC = `/assets/img/${id}.jpg`
  const infoImagePath = `/assets/svg/info.svg`
  return (

    <div className='col animate__animated animate__fadeIn'>
      
      <div className='card p-3' style={{ width: 18 + 'rem'}}>  
        <img
          src={imageSRC}
          className='card-img-top img-thumbnail'
          alt={superhero}
        />

        <div className='card-body d-flex flex-column justify-content-center align-items-center'>
          <h5 className="card-title"> {superhero} </h5>
          <p className="card-text"> {publisher} </p>
          <div className='d-flex flex-row '>
            <p className="card-text p-2"> {alter_ego} </p>
            <p className="card-text p-2">
              {
                ( alter_ego !== characters ) && 
                  <small className='text-muted'> {characters} </small>
              }
            </p>
          </div>
          <p className="card-text p-2"> {first_appearance} </p>
        </div>
        
        <Link to={`/hero/${id}`}>
          <button
            className='btn btn-dark w-100'
            type='button'
          >
            Mas informacion <img src={infoImagePath} alt='info' />
          </button>
        </Link>
      </div>
    </div>
  )
};
