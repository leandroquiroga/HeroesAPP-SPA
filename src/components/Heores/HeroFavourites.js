import React from 'react'
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';



export const HeroFavourites = (props) => {
  const { data, pathCloseSVG, dispatchHero } = props;

  const { id, superhero, publisher, characters, first_appearance } = data;

  const handleDeleteHero = ((heroID) => {
  
    
    dispatchHero({
      type: 'deleteHero',
      payload: heroID
    });

    
    (function notify() {
      toast.error('Se ha eliminado un heroe', {
        theme: 'colored',
        position: "top-right",
        autoClose: 2000,
        hideProgressBar: false,
        closeOnClick: true,
        pauseOnHover: true,
        draggable: true,
        progress: undefined,
      });
    })();
  });

  return (
    <div
      className='col animate__animated animate__fadeIn px-2'
    >
      <li
        className='card'
      >
        <div className='d-flex justify-content-end align-items-center'>
          <button
            className='main-container--button'
          >
            <img
              src={pathCloseSVG}
              alt='delete'
              className='main-container--button-close'
              onClick={() => handleDeleteHero(id)}
            />
          </button>
          <ToastContainer />
        </div>
        <img 
          src={`/assets/img/${id}.jpg`}
          alt={superhero}
          className='card-img-top img-thumbnail'
        />
        
        <div className='card-body d-flex flex-column justify-content-center align-items-center'>
          <h5 className="card-title"> {superhero} </h5>
          <p className="card-text"> {publisher} </p>
          <div className='d-flex flex-row '>
            <p className="card-text p-2">
                  <small className='text-muted'> {characters} </small>
            </p>
          </div>
          <p className="card-text p-2"> {first_appearance} </p>
        </div>
      </li>
    </div>
  )
}
