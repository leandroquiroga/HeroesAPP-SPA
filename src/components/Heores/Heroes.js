import React, { useMemo } from 'react';
import { useParams, useNavigate, Navigate } from 'react-router-dom';
import { getHeroById } from './../../helpers/index';
import { ToastContainer, toast } from 'react-toastify';

import 'react-toastify/dist/ReactToastify.css';
export const Heroes = () => {
  // const [msgAdd , setMsgAdd] = useState(false)  
  const { heroID } = useParams();
  const navigate = useNavigate();
  const hero = useMemo(() => getHeroById(heroID), [heroID] ) 

  const { id, superhero, alter_ego, characters, first_appearance, publisher } = hero; 

  if (!hero) {
    return <Navigate to='/' />
  };

  // Regresa al url anterior de la pila del stack
  const handleBack = () => navigate(-1);


  const notify = () => {
    toast.success('Nuevo heroe agregado', {
      theme: 'colored',
      position: "bottom-right",
      autoClose: 2000,
      hideProgressBar: false,
      closeOnClick: true,
      pauseOnHover: true,
      draggable: true,
      progress: undefined,
      });
  }
  

  const imagePath = `/assets/img/${id}.jpg`
  const startImagePath = `/assets/svg/start.svg`
  const backImagePath = `/assets/svg/back.svg`
  return (
    <div className='container animate__animated animate__fadeIn '>
      <div className='d-flex flex-column mt-3 flex-md-row '>
        <div className='col-12 col-md-4 '>
          <img
            src={imagePath}
            alt={superhero}
            className='img-fluid w-100'
          >
          </img>
        </div>
        <div className='col-12 col-md-8 p-5'>
            <h3>{superhero}</h3>
            <ul className='list-group list-group-flush'>
              <li className='list-group-item'>
                <b>Alter ego: </b>
                  {alter_ego}
              </li>
              <li className='list-group-item'>
                <b>Publisher: </b>
                  {publisher}
              </li>
              <li className='list-group-item'>
                <b>First Appearance: </b>
                  {first_appearance}
              </li>
            </ul>

            <h5 className='mt-3 mx-3'> Characters </h5>
            <p className='mx-3'> {characters} </p>

          <div className='d-flex flex-row '>
              <button
                className='btn btn-danger m-2'
                onClick={handleBack}
              >
                Regresar <img src={backImagePath} alt='favorite' />
              </button>
              <div className='d-flex flex-column flex-md-row align-items-center'>
                <button
                  className='btn btn-danger m-2'
                  onClick={notify}>
                  AÃ±adir a <img src={startImagePath} alt='favorite' />
              </button>
              <ToastContainer />
            </div>  
          </div>

        </div>
      </div>
    </div>
  );
};
