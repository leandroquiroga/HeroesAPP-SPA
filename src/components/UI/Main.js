import React from 'react';

export const Main = () => {
  const pathIMG = `/assets/img/not-super-hero.png`
  return (
       
    <div className=' container d-flex justyfy-content-center align-items-center'>

    {/*
        Si hay elemento en arreglo de favoritos,
        muestra una lista de todos los elementos 
    
     */}

    <div className='d-flex justify-content-between align-items-center w-100 p-5 flex-column flex-lg-row '>
      <img 
        className='img-fluid'
        alt='have not element'
        src={pathIMG}
      />

      <p className='text-center fs-3 mt-5 mt-lg-0'>
        No hay superheroes en la lista de favoritos...
      </p>
    </div>
  </div>
  );
};
