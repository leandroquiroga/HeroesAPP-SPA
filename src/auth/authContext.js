import { createContext, useEffect, useReducer, useState } from "react";
import { signInWithEmailAndPassword, signOut ,createUserWithEmailAndPassword } from 'firebase/auth';
import { authReducer } from "./authReducer";
import { auth } from './../firebase';
import { heroReducer } from "../helpers";

// Retorna el usuario si lo encuentra en localStorage si no retorna el logged en false
const init = () => JSON.parse(localStorage.getItem('user')) || {logged: false}

// Retorna los hero si lo encuntra en localStorage si no retorna un array 
const initHero = () => JSON.parse(localStorage.getItem('hero')) || [];

export const AuthContext = createContext();

export const AuthProvaider = (props) => {
  const [user, dispatch] = useReducer(authReducer, {}, init);
  const [currentUser, setCurrentUser] = useState({});
  const [hero, dispatchHero] = useReducer(heroReducer, [], initHero);

  // Cuando inicia la aplicacion va a estar observando si hay un cambio
  // de usuario.
  useEffect(() => {
    auth.onAuthStateChanged((user) => setCurrentUser(user))
  }, []);

  // Esta pendiente al user para ir almancendolo en el localStorage
  useEffect(() => {
    localStorage.setItem('user', JSON.stringify(user))
    return () => {
      if (!user) return;
    }
  }, [user]);

  // Esta pendiente al hero, cuando detecta un cambio lo almacena en el localStorage
  useEffect(() => {
    localStorage.setItem('hero', JSON.stringify(hero));
    return () => {
      if (!hero) return;
    }
  }, [hero])
  


  const singUp = (email, password) => createUserWithEmailAndPassword(auth, email, password);

  const logIn = (email, password) => signInWithEmailAndPassword(auth, email, password);
  
  const logOut = (auth) => signOut(auth);

  const value = {
    user,
    dispatch,
    singUp,
    logIn,
    logOut,
    currentUser,
    dispatchHero,
    hero,
  }
  return (
    <AuthContext.Provider value={value}>
      {props.children}
    </AuthContext.Provider>
  )
}