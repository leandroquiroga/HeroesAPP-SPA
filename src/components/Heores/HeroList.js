import React, { useMemo } from 'react';
import { getHeroByPublisher } from '../../helpers';
import { HeroCard } from './HeroCard';


export const HeroList = ({publisher}) => {
  
  const heroes = useMemo(() => getHeroByPublisher(publisher), [publisher]);

  return (
    <div className=' animate__animated animate__fadeIn row justify-content-center row-cols-1 row-cols-md-2  row-cols-xxl-3 g-2'>
      
      {
        heroes.map(hero => ( 
          <HeroCard
            key={hero.id}
            {...hero}   
          />
        ))
      }
      
    </div>
  );
};