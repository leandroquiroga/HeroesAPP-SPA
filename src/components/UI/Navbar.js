import React, { useContext } from 'react';
import { Link, NavLink,  useNavigate } from 'react-router-dom';
import { AuthContext } from './../../auth/autoContext';
import { types } from './../../types/type';

export const NavbarScreen = () => {

  const navigate = useNavigate();
  const { user, dispatch } = useContext(AuthContext)
  // retorna a inicio de sesion

  const handleLoguot = () => {
    const action = {
      type: types.logout,
    };
  
    dispatch(action);
    navigate('/login', {
      replace: true
    })
  };
  return (
    <nav className='navbar navbar-expand-lg navbar-dark bg-dark px-2'>

      <Link
        className='navbar-brand'
        to="/"
      >
        Inicio
      </Link>
      <div className='navbar-collapse'>
        <div className='navbar-nav'>
          <NavLink
            className={({ isActive }) => 'nav-item nav-link text-white' + (isActive ? 'active' : '' )}
            to="/marvel"
          >
            Marvel
          </NavLink>

          <NavLink
            className={({ isActive }) => 'nav-item nav-link text-white' + (isActive ? 'active' : '' )}
            to='/dc'
          >
            DC
          </NavLink>
          <NavLink
            className={({ isActive }) => 'nav-item nav-link text-white' + (isActive ? 'active' : '' )}
            to='/search'
          >
            Search
          </NavLink>
      </div>
    </div>
    <div className='navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end'>
      <ul className='navbar-nav ml-auto'>

        <span className='nav-item nav-link text-white'>
          Hola, {user.name} !
        </span>

        <button
          className='nav-item btn btn-danger'
          onClick={handleLoguot}
        >
          Cerrar sesi√≥n
        </button>
      </ul>
    </div>
  </nav>
  );
}