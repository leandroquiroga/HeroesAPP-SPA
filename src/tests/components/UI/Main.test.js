import { mount, shallow } from "enzyme";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { AuthContext } from "../../../auth/authContext";
import { HeroFavourites } from "../../../components/Heores/HeroFavourites";
import { Main } from "../../../components/UI/Main";

describe('Testing on component <Main />', () => {

  test('This test must showing the component <Main /> correctly', () => {
    const contextValueHero = {
      dispatch: jest.fn(),
      hero: [
        {
          'id': 'dc-batman',
          'superhero': 'Batman',
          'publisher': 'DC Comics',
          'alter_ego': 'Bruce Wayne',
          'first_appearance': 'Detective Comics #27',
          'characters': 'Bruce Wayne'
        },
      ]
    };
  
    const wapper = mount(
      <AuthContext.Provider value={contextValueHero}>
        <MemoryRouter initialEntries={['/']}>
          <Routes>
            <Route path="/" element={<Main />} />
          </Routes>
        </MemoryRouter>
      </AuthContext.Provider>
    )
    
    expect(wapper).toMatchSnapshot();
    expect(wapper.find('.container-card').exists()).toBe(true);
    expect(wapper.find('h5').text().trim()).toBe('Batman');
    expect(wapper.find('.Toastify').exists()).toBe(true)

  });




  test('This test must showing the componente <Main /> if hero.length === 0', () => {
    const contextValueHero = {
      dispatch: jest.fn(),
      hero: []
    };

  const pathIMG = `/assets/img/not-super-hero.png`


    const wapper = mount(
      <AuthContext.Provider value={contextValueHero}>
        <MemoryRouter initialEntries={['/']}>
          <Routes>
            <Route path="/" element={<Main />} />
          </Routes>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wapper).toMatchSnapshot();
    expect(wapper.find('p').text().trim()).toBe('No hay superheroes en la lista de favoritos...')
    expect(wapper.find('img').prop('src')).toBe(pathIMG);
  });
});

describe('Testing on component <HeroFavourites />', () => {
  const contextValue = {
    dispatch: jest.fn(),
    hero: [
      {
        'id': 'dc-batman',
        'superhero': 'Batman',
        'publisher': 'DC Comics',
        'alter_ego': 'Bruce Wayne',
        'first_appearance': 'Detective Comics #27',
        'characters': 'Bruce Wayne'
      },
    ]
  };
  const { hero, dispatch } = contextValue;
  const pathCloseSVG = `/assets/svg/close.svg`

  const wapperHero = shallow(
    <HeroFavourites
      key={hero.id}
      data={hero}
      pathCloseSVG={pathCloseSVG}
      dispatchHero={dispatch}
    />
  );
  test('This test must showing the comoponente <HeroFavourites /> with their props respectively', () => {

    expect(wapperHero).toMatchSnapshot();
  });
})