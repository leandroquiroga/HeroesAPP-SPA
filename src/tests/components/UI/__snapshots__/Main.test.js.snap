// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on component <HeroFavourites /> This test must showing the comoponente <HeroFavourites /> with their props respectively 1`] = `
<div
  className="col animate__animated animate__fadeIn px-2"
>
  <li
    className="card"
  >
    <div
      className="d-flex justify-content-end align-items-center"
    >
      <button
        className="main-container--button"
      >
        <img
          alt="delete"
          className="main-container--button-close"
          onClick={[Function]}
          src="/assets/svg/close.svg"
        />
      </button>
      <ToastContainer
        autoClose={5000}
        closeButton={[Function]}
        closeOnClick={true}
        draggable={true}
        draggableDirection="x"
        draggablePercent={80}
        hideProgressBar={false}
        newestOnTop={false}
        pauseOnFocusLoss={true}
        pauseOnHover={true}
        position="top-right"
        role="alert"
        rtl={false}
        theme="light"
        transition={[Function]}
      />
    </div>
    <img
      className="card-img-top img-thumbnail"
      src="/assets/img/undefined.jpg"
    />
    <div
      className="card-body d-flex flex-column justify-content-center align-items-center"
    >
      <h5
        className="card-title"
      >
         
         
      </h5>
      <p
        className="card-text"
      >
         
         
      </p>
      <div
        className="d-flex flex-row "
      >
        <p
          className="card-text p-2"
        >
          <small
            className="text-muted"
          >
             
             
          </small>
        </p>
      </div>
      <p
        className="card-text p-2"
      >
         
         
      </p>
    </div>
  </li>
</div>
`;

exports[`Testing on component <Main /> This test must showing the component <Main /> correctly 1`] = `
<main
  className="container p-4"
>
  <div>
    <h1
      className="text-center"
    >
       Superheroes favoritos 
    </h1>
    <hr />
    <div
      className="container-card"
    >
      <div
        className="col animate__animated animate__fadeIn px-2"
      >
        <li
          className="card"
        >
          <div
            className="d-flex justify-content-end align-items-center"
          >
            <button
              className="main-container--button"
            >
              <img
                alt="delete"
                className="main-container--button-close"
                onClick={[Function]}
                src="/assets/svg/close.svg"
              />
            </button>
            <div
              className="Toastify"
            />
          </div>
          <img
            alt="Batman"
            className="card-img-top img-thumbnail"
            src="/assets/img/dc-batman.jpg"
          />
          <div
            className="card-body d-flex flex-column justify-content-center align-items-center"
          >
            <h5
              className="card-title"
            >
               
              Batman
               
            </h5>
            <p
              className="card-text"
            >
               
              DC Comics
               
            </p>
            <div
              className="d-flex flex-row "
            >
              <p
                className="card-text p-2"
              >
                <small
                  className="text-muted"
                >
                   
                  Bruce Wayne
                   
                </small>
              </p>
            </div>
            <p
              className="card-text p-2"
            >
               
              Detective Comics #27
               
            </p>
          </div>
        </li>
      </div>
    </div>
  </div>
</main>
`;

exports[`Testing on component <Main /> This test must showing the componente <Main /> if hero.length === 0 1`] = `
<main
  className="container p-4"
>
  <div
    className="d-flex justify-content-between align-items-center w-100 p-5 flex-column flex-lg-row "
  >
    <img
      alt="have not element"
      className="img-fluid"
      src="/assets/img/not-super-hero.png"
    />
    <p
      className="text-center fs-3 mt-5 mt-lg-0"
    >
      No hay superheroes en la lista de favoritos...
    </p>
  </div>
</main>
`;
